### ISSUE TITLE
flux diff ks doesn't list removed resources

### ISSUE THREAD

**Issue by tun0** [fluxcd/flux2#3733]
### Describe the bug

`flux diff ks` only lists drifted and created resources. I'd expect removed resources to be listed as well. Optionally with configurable behavior regarding `.spec.prune` (warn, ignore, etc).

### Steps to reproduce

1. On an existing install, remove a resource (either by removing a file or removing a reference from `kustomization.yaml`).
2. Run `flux diff ks ...`
3. No changes are listed

### Expected behavior

The removed resource(s) to be listed as such.

### OS / Distro

N/A

### Flux version

v0.41.2

### Flux check

► checking prerequisites
✔ Kubernetes 1.24.9-gke.3200 >=1.20.6-0
► checking controllers
✔ helm-controller: deployment ready
► ghcr.io/fluxcd/helm-controller:v0.31.2
✔ image-automation-controller: deployment ready
► ghcr.io/fluxcd/image-automation-controller:v0.31.0
✔ image-reflector-controller: deployment ready
► ghcr.io/fluxcd/image-reflector-controller:v0.26.1
✔ kustomize-controller: deployment ready
► ghcr.io/fluxcd/kustomize-controller:v0.35.1
✔ notification-controller: deployment ready
► ghcr.io/fluxcd/notification-controller:v0.33.0
✔ source-controller: deployment ready
► ghcr.io/fluxcd/source-controller:v0.36.1
► checking crds
✔ alerts.notification.toolkit.fluxcd.io/v1beta2
✔ buckets.source.toolkit.fluxcd.io/v1beta2
✔ gitrepositories.source.toolkit.fluxcd.io/v1beta2
✔ helmcharts.source.toolkit.fluxcd.io/v1beta2
✔ helmreleases.helm.toolkit.fluxcd.io/v2beta1
✔ helmrepositories.source.toolkit.fluxcd.io/v1beta2
✔ imagepolicies.image.toolkit.fluxcd.io/v1beta2
✔ imagerepositories.image.toolkit.fluxcd.io/v1beta2
✔ imageupdateautomations.image.toolkit.fluxcd.io/v1beta1
✔ kustomizations.kustomize.toolkit.fluxcd.io/v1beta2
✔ ocirepositories.source.toolkit.fluxcd.io/v1beta2
✔ providers.notification.toolkit.fluxcd.io/v1beta2
✔ receivers.notification.toolkit.fluxcd.io/v1beta2
✔ all checks passed

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by souleb** [comment]
Hello @tun0, can you give us an example, so we can reproduce this?

**Comment by tun0** [comment]
Hi @souleb, apparently that's a very good question. When I try to reproduce it I can't. I'm gonna try 'n figure out what the exact scenario was that led me to create this issue!

**Comment by tun0** [comment]
I figured out what caused us to think deletes weren't supported. And it's a tricky one: we're using `--kustomization-file`, so there's no list of resources to check against.

Possible solution: when `--kustomization-file` is specified (optionally) look at the in-cluster Kustomization as well (specifically for the list managed resources). This might also open the door for another use-case: showing a diff for a deleted Kustomization (which might become a bit more tricky).

**Comment by aldojimenez** [comment]
We are in the same situation, we have a CI pipeline in our fluxcd gitops repo to perform these diff checks. We use the `--kustomization-file` to ensure we account for modifications to the `Kustomization` resource not just the path.

**Comment by BlackDark** [comment]
Maybe related to this issue. For me the diff works the outputs contains the deleted resources but the exit code is not as expected 1 instead it is just 0.

```bash
flux diff kustomization permissions --namespace mynamespace --path flux/releases/permissions/overlays/dev
✓  Kustomization diffing...
► RoleBinding/mynamespace/myrole-admin-test deleted

echo $?
0 # imho should be 1 here
```

**Comment by souleb** [comment]
@BlackDark can you open a different issue for this please?

**Comment by williambrode** [comment]
Is it possible to do what @tun0 suggested?

> Possible solution: when --kustomization-file is specified (optionally) look at the in-cluster Kustomization as well (specifically for the list managed resources). This might also open the door for another use-case: showing a diff for a deleted Kustomization (which might become a bit more tricky).

I don't know the code but this is just how I would expect it to work with the `--kustomization-file` option.  Feels more like a bug than an intentional side affect of this option?

Recap: Using `--kustomization-file` (which we should do to ensure changes to the flux kustomization are realized) doesn't show deleted objects.  Pretty bad false positive since we may submit changes thinking they are great and find out we accidentally deleted everything 😮.  It would likely be better to simply remove this option until the bug can be fixed.

**Comment by souleb** [comment]
With `flux diff ks`, once we have retrieved a `kustomization` through a  `--kustomization-file` flag or by querying the `kube-api server`, we build the target list of `kubernetes objects` and apply the needed transformations . Then, we ask the `kube-api server` to perform the `diff` (with server side apply) against what's currently running in cluster. The behaviour is the same in both cases.

> Possible solution: when --kustomization-file is specified (optionally) look at the in-cluster Kustomization as well (specifically for the list managed resources). 

I think that if there is already a `kustomization` object in the cluster, we expect it to be successfully reconciled. So we don't need to query it, because when we ask the `kube-api server` to perform the `diff`, it should see all the `kustomization` managed resources.

**Comment by williambrode** [comment]
This seems like the problematic part:

> With flux diff ks, once we have retrieved a kustomization through a --kustomization-file flag or by querying the kube-api server

Since when we pass in `--kustomization-file` we need to retrieve the kustomization from BOTH the kube-api and the local file.  Then we need to look for missing or added resources before proceeding to diff only the resources in the local kustomization file.

**Comment by souleb** [comment]
Maybe I'm unclear, that' not what I meant. We don't do both. With `--kustomization-file` we retrieve the file locally, resolve the path to build locally, then perform the diff.

**Comment by williambrode** [comment]
Yes, I understood what you were saying.  But we NEED to do both, right?  Otherwise we do not report removed resources.  That is the bug we are seeing.

**Comment by souleb** [comment]
I think that's where I would need an example of what you observe.

**Comment by williambrode** [comment]
Here is the Kustomization i was using:

```
apiVersion: kustomize.toolkit.fluxcd.io/v1beta1
kind: Kustomization
metadata:
  name: foo
  namespace: flux-system
spec:
  interval: 2m0s
  sourceRef:
    kind: GitRepository
    name: flux-system
  path: ./foo
  prune: true
```

With files like this with resources in them:

```
./foo/bar.yaml
./foo/baz.yaml
```

Apply that to the cluster, then delete `./foo/baz.yaml`.  Run `flux diff kustomization foo --path ./foo --kustomization-file foo.yaml` and notice that it will not report any differences.

**Comment by stefanprodan** [comment]
@williambrode the `v1beta1` API didn’t support diffing. Can you post here `flux check` please.
